### DISCUSSION TITLE
Test suite errors/failures, some fixes, observations, etc

### DISCUSSION THREAD

**Discussion by silnarm** [livewire/livewire#4590] [7. Bugs (archived)] [answered]
Hi,

I'm still fairly new to both Livewire and Laravel, but have been building something with it for a couple of months now and wanted to look into a few issues stemming from that, following the contrib guide has led me on an interesting diversion with the test suite instead :-)

I think maybe some kind of setup issue, but on failure/error I was getting this:
```
[jmcculloch@jupiter livewire-git]$ phpunit --filter it_shows_page_expired_dialog_when_livewire_deployment_invalidation_hash_has_changed

	==== Redirecting to composer installed version in vendor/phpunit ====

	PHPUnit 9.5.13 by Sebastian Bergmann and contributors.

	Psy Shell v0.11.1 (PHP 8.0.15 — cli) by Justin Hileman
	-->  dusk
	RuntimeException with message 'Too many arguments, expected arguments "command".'
	>>> 
```

It looks like once in the tinker shell it was going to give me stack trace, but something else is erroring out beforehand. I don't know much about tinker and have no idea what is going on here, if anyone can enlighten me I would be grateful.
I got around this by commenting out the breakIntoATinkerShell() lines in Tests\Browser\TestCase.

### Tests\Browser\Deployment\Test::it_shows_page_expired_dialog_when_livewire_page_has_expired_exception_is_manually_thrown

These tests error for me, the pause() is apparently not long enough, and assertDialogOpened() is assuming a dialog is up, and only asserts on the text, so errors out rather than failing, possibly worth a dusk bug report.
It is not a good idea to bake fixed delays into automated tests, though unavoidable in many cases.
Here though there is no need, we can just wait for an alert, default (no param) is 5 a second time-out which while extreme seems fine to me.
So insteand of ->pause(50) we just ->waitForDialog()
[This this is implemented as a proper 'wait' on the browser driver, it polls for the condition and this will not cause a 5 second delay in the test]

### Tests\Browser\Loading\Test::test_different_delay_durations

This one was failing consistently for me, but in different places on different runs. Always the first of the not-visible checks in one of the 'sub' tests though, and generally happended on the shorter or short (sub)tests.

My first thoughts were that the pause() times were _very_ tight, and that a pause half way between the delay we are testing and the next one would make more sense. Of course this doesn't make sense with the diagnostics I just reported, and indeed just exacerbates the problem.

Some further, and apparently better, thought suggests what the problem must really be. Things take time. By the time we are checking for the first thing we expect not to see, it can be seen and correctly so, because more time has passed while we were checking other things.

If I replace test_different_delay_durations() with this:

```
    private function assertVisibleAndNotVisible(Browser $browser, array $visible, array $invisible) {
        echo __METHOD__ . " visible count = " . count($visible) . PHP_EOL;
        $t = -microtime(true);
        foreach ($visible as $item) {
            $browser->assertVisible($item);
        }
        $t += microtime(true);
        $t = intval($t * 1000);
        echo __METHOD__ . " asserting visible components took {$t} ms" . PHP_EOL;
        foreach ($invisible as $item) {
            $browser->assertNotVisible($item);
        }
    }

    public function test_different_delay_durations() {
        $this->browse(function ($browser) {
            Livewire::visit($browser, ComponentWithLoadingDelays::class);
            $this->assertVisibleAndNotVisible($browser, [], ['@delay-shortest', '@delay-shorter', '@delay-short', '@delay', '@delay-long', '@delay-longer', '@delay-longest']);
            $browser->waitForLivewire(function (Browser $browser) {
                $browser->click('@load');
                $browser->pause(51);
                $this->assertVisibleAndNotVisible($browser, ['@delay-shortest'], ['@delay-shorter', '@delay-short', '@delay', '@delay-long', '@delay-longer', '@delay-longest']);
            })->waitForLivewire(function (Browser $browser) {
                $browser->click('@load');
                $browser->pause(101);
                $this->assertVisibleAndNotVisible($browser, ['@delay-shortest', '@delay-shorter'], ['@delay-short', '@delay', '@delay-long', '@delay-longer', '@delay-longest']);
            })->waitForLivewire(function (Browser $browser) {
                $browser->click('@load');
                $browser->pause(151);
                $this->assertVisibleAndNotVisible($browser, ['@delay-shortest', '@delay-shorter', '@delay-short'], ['@delay', '@delay-long', '@delay-longer', '@delay-longest']);
            })->waitForLivewire(function (Browser $browser) {
                $browser->click('@load');
                $browser->pause(201);
                $this->assertVisibleAndNotVisible($browser, ['@delay-shortest', '@delay-shorter', '@delay-short', '@delay'], ['@delay-long', '@delay-longer', '@delay-longest']);
            })->waitForLivewire(function (Browser $browser) {
                $browser->click('@load');
                $browser->pause(301);
                $this->assertVisibleAndNotVisible($browser, ['@delay-shortest', '@delay-shorter', '@delay-short', '@delay', '@delay-long'], ['@delay-longer', '@delay-longest']);
            })->waitForLivewire(function (Browser $browser) {
                $browser->click('@load');
                $browser->pause(501);
                $this->assertVisibleAndNotVisible($browser, ['@delay-shortest', '@delay-shorter', '@delay-short', '@delay', '@delay-long', '@delay-longer'], ['@delay-longest']);
            })->waitForLivewire(function (Browser $browser) {
                $browser->click('@load');
                $browser->pause(1001);
                $this->assertVisibleAndNotVisible($browser, ['@delay-shortest', '@delay-shorter', '@delay-short', '@delay', '@delay-long', '@delay-longer', '@delay-longest'], []);
            });
        });
    }
```

Then I get results that look like this:

```
[jmcculloch@jupiter livewire-git]$ phpunit --filter different_delay_durations

==== Redirecting to composer installed version in vendor/phpunit ====

PHPUnit 9.5.13 by Sebastian Bergmann and contributors.

F                                                                   1 / 1 (100%)Tests\Browser\Loading\Test::assertVisibleAndNotVisible visible count = 0
Tests\Browser\Loading\Test::assertVisibleAndNotVisible asserting visible components took 0 ms
Tests\Browser\Loading\Test::assertVisibleAndNotVisible visible count = 1
Tests\Browser\Loading\Test::assertVisibleAndNotVisible asserting visible components took 20 ms
Tests\Browser\Loading\Test::assertVisibleAndNotVisible visible count = 2
Tests\Browser\Loading\Test::assertVisibleAndNotVisible asserting visible components took 29 ms
Tests\Browser\Loading\Test::assertVisibleAndNotVisible visible count = 3
Tests\Browser\Loading\Test::assertVisibleAndNotVisible asserting visible components took 97 ms


Time: 00:04.076, Memory: 22.00 MB

There was 1 failure:

1) Tests\Browser\Loading\Test::test_different_delay_durations
Element [body [dusk="delay"]] is visible.
Failed asserting that true is false.
```
So on the first subtest we took 20ms asserting visible things, the difference in delay to the next is 50, so we had 30ms to play with and all was well.
On the second subtest we took 29ms on the visibles, delay difference to first that shouldn't be shown is 50 again, so 21ms to play with and all is well.
Third subtest was 97ms checking visible elements, delay difference is 50 again here, and we are already 47ms overtime before we even start the check.

Potential rabbit hole, I don't think we would want to be adding the timekeeping code to make sensible assertions, and that would mean conditional assertions which just seems very dodgy.
So I modified it to just check visibility of the one we are expecting to have just become visible and the next one we are expecting to not be visible (yet).

```
F                                                                   1 / 1 (100%)Tests\Browser\Loading\Test::assertVisibleAndNotVisible asserting visible element took 0 ms
Tests\Browser\Loading\Test::assertVisibleAndNotVisible asserting visible element took 19 ms
Tests\Browser\Loading\Test::assertVisibleAndNotVisible asserting visible element took 41 ms
```
Still fails, in this example the (single element) visible assertion took 41ms (subtest 2), leaving 9ms which apparently wasn't enough to get to the 'meat' of the assertNotVisible().

So I swapped the tests, assert the not visible one first, then the visible one. This works consistently, but we are of course now delaying the visible check.
Given the nature of the tests, and the time sensitivity, I am now thinking the best thing is to actually make these separate (sub)tests, with each asserting just one visible/not-visible state on one element at the delay. The underlying issue remains of course, and if you wanted to add a test to assert non visibility of something you are expecting to become visible very soon you would have issues with it again, but for the existing tests it should get rid of the false failures.

Not ideal, but probably nor is it worth going to a lot of trouble with. Thoughts?

Other possibilites / Things potentially worth exploring:
 * is my chromedriver or setup broken? While not new my computer is also not really old (8th gen i7),
    are the times I have reported here reasonable?
 * is finding the elements to assert visibility of by a custom 'dusk' attribute hurting?
 * would it be better to implement the tests entirely within an ->executeScript() call?
 

### Other intermitent failures/errors?

I had some failures very early on in the dusk tests when I first started running the tests, because of the other problems I had I never got to investigate as they appear to have not persisted (I believe they were earlier in the suite than Deployment tests, but could be wrong and it could have been them, I've lost all that console history now).

I also had a failure on the Polling (unnamed) test when running the full suite (with the other tests fixed as above), but naming it and targetting with --filter just gives me pass after pass.
This was the full suite output:
```
[jmcculloch@jupiter livewire-git]$ phpunit

==== Redirecting to composer installed version in vendor/phpunit ====

PHPUnit 9.5.13 by Sebastian Bergmann and contributors.

...............................................................  63 / 620 ( 10%)
............................................................... 126 / 620 ( 20%)
............................................................... 189 / 620 ( 30%)
............................................................... 252 / 620 ( 40%)
............................................................... 315 / 620 ( 50%)
............................................................... 378 / 620 ( 60%)
.........................SS.................................... 441 / 620 ( 71%)
............................................................... 504 / 620 ( 81%)
...................S........................................... 567 / 620 ( 91%)
.........F...........................W...............           620 / 620 (100%)

Time: 03:54.675, Memory: 68.00 MB

There was 1 warning:

1) Warning
No tests found in class "Tests\Browser\Stacks\Test".

--

There was 1 failure:

1) Tests\Browser\Polling\Test::test
Did not see expected text [6] within element [body [dusk="output"]].
Failed asserting that false is true.

/home/jmcculloch/livewire-git/vendor/laravel/dusk/src/Concerns/MakesAssertions.php:181
/home/jmcculloch/livewire-git/tests/Browser/Polling/Test.php:40
/home/jmcculloch/livewire-git/tests/Browser/TestCase.php:254
/home/jmcculloch/livewire-git/vendor/laravel/dusk/src/Concerns/ProvidesBrowser.php:69
/home/jmcculloch/livewire-git/tests/Browser/TestCase.php:264
/home/jmcculloch/livewire-git/tests/Browser/Polling/Test.php:50

FAILURES!
Tests: 620, Assertions: 3719, Failures: 1, Warnings: 1, Skipped: 3.
```

I have noticed an aweful lot of pause() calls in there. Many may be unavoidable, but I'm sure there are a lot that should be waits and in their current form may generate false failures sometimes.
I started tinkering with the source to look at something else, but I'm learning a great deal about livewire very quickly playing with the tests, is converting pauses to waits worth doing where we can?

**Comment by joshhanley** [comment, marked as answer]
@silnarm yeah some of our dusk tests can be a bit flaky at times (especially when running in GitHub actions), so we're definitely open to possible solutions to improve that/ fix permanently.

But, that being said, whenever I run them locally, I nearly never have a failure, and that's one thing we ensure is that all tests are running and pass locally and also a few tests suites (if not all) pass on GH actions before we tag a release. So it hasn't really been an issue worth spending time on when there are other bugs we should be looking at first.

For example, this is a run I just did from the master branch 

![image](https://user-images.githubusercontent.com/882837/152729657-1a194e5b-1c8e-4f05-9667-0dbe0f45baa6.png)

I did review the loading test recently and got it to fail less and managed to fix a few others, but eventually left it as good enough haha.

Typically for me, the only test that may fail is that loading test, so seems unusual that those other tests are failing for you.

We do welcome PR's to improve things though, so if you do have some fixes worth submitting, please feel free to submit them in seperate PR's 🙂

**Reply by silnarm** [reply]
Cheers.

I think from the sounds of things there probably is something amiss in my setup, or I have a faulty webdriver. The time it is taking for assertNotVisible (for me) does seem excessive, and the tests in test_different_delay_durations() were always going to be a problem with such delays. The other failures I reported (that were consistent and I was able to investigate) are also essentially timing related, so quite possibly have the same root cause.

Interesting about the loading tests, there were a couple of failures I got early, while the break into tinker bug was preventing any meaningful output, it may have been amongst them. I have moved on from the test suite, but imagine I will be running it again this weekend, so I will keep an eye out.

**Reply by joshhanley** [reply]
@silnarm no worries! What setup are you using (OS, valet/docker/etc)?

Yep due to the nature of Livewire we do have quite a few timing based tests, which can be fickle at times 😕

Let me know how you go!

**Reply by silnarm** [reply]
Linux, with everything native (no containers or VMs).

Didn't get to the full suite last weekend, was too busy trying to make new tests I added not fail to be running others :-) Probably no chance this weekend, but I think running them on another computer and doing the timing there might be more revealing anyway (or perhaps I should just try it again after rebooting this one!). I'll try to get it going on my work computer (Windows) in "My 10%" time next week.
